pkg/chaosctl/recover/networkchaos.go

```diff
@@ -0,0 +1,183 @@
+// Copyright 2021 Chaos Mesh Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package recover
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/hasura/go-graphql-client"
+	"github.com/pingcap/errors"
+
+	ctrlclient "github.com/chaos-mesh/chaos-mesh/pkg/ctrl/client"
+)
+
+// match handle number and device in tc qdisc rules
+// ### Rules Example:
+// We can match handle number `8001` and device `vnet0` in the following rule:
+// ```
+// qdisc noqueue 8001: dev vnet0 root refcnt 2
+// ```
+var tcRegexp = regexp.MustCompile(`([0-9]+): dev (\w+)`)
+
+type tcsRecoverer struct {
+	client *ctrlclient.CtrlClient
+}
+
+func newTcsRecoverer(client *ctrlclient.CtrlClient) Recoverer {
+	return &tcsRecoverer{
+		client: client,
+	}
+}
+
+// Recover clean all tc qdisc rules
+func (r *tcsRecoverer) Recover(ctx context.Context, pod *PartialPod) error {
+	deviceSet := map[string]bool{}
+	for _, rule := range pod.TcQdisc {
+		matches := tcRegexp.FindStringSubmatch(rule)
+		if len(matches) != 3 {
+			continue
+		}
+		handle, err := strconv.Atoi(matches[1])
+		if err != nil {
+			return errors.Wrapf(err, "parse tc qdisc handle: `%s`", matches[1])
+		}
+		if handle > 0 {
+			// if handle is 0, it's unnecessary to clean
+			deviceSet[matches[2]] = true
+		}
+	}
+
+	var devices []graphql.String
+	for dev := range deviceSet {
+		devices = append(devices, graphql.String(dev))
+	}
+
+	if len(devices) == 0 {
+		printStep("all tc rules are cleaned up")
+		return nil
+	}
+	printStep(fmt.Sprintf("cleaning tc rules for device %v", devices))
+
+	var mutation struct {
+		Pod struct {
+			CleanTcs []string `graphql:"cleanTcs(devices: $devices)"`
+		} `graphql:"pod(ns: $ns, name: $name)"`
+	}
```

Comment:
I am curious about why should we define this struct called `mutation` here.

Could it be a part of `ctrl/server` API?

Reply:
Similar question relates to https://github.com/chaos-mesh/chaos-mesh/pull/2992/files#r858287149

It makes me confusing;;

Reply:
This is a structure to generate a request and map the response. It's a calling of API https://github.com/Hexilee/chaos-mesh/blob/chaosctl-force-recover/pkg/ctrl/server/schema.graphqls#L135.

The mutation is the root name of all mutable APIs.

Reply:
I think I could guess the usage of that `mutation`. I just consider why do we NOT put it as a part of `ctrl/client`;

> Sorry I just incorrectly used `ctrl/server` in the previous messages;

Reply:
Because mutations of graphql are rarely reusable, in future, we may clean tcs rules and iptables rules in single mutation.

We can merge multiple queries like the `PartialPod` in `ctrl/client`, but we cannot merge mutations.

Reply:
IIUC, do you mean the current `mutation` graphql API is not enough stable now, so we prefer to use it as an "internal API"? 

Or there are some restrict of graphql, which makes it not so easy to expose it as public and stable API in `ctrl/client`?

Reply:
It's irrelevant about stability, it means dynamic API subset in single calling.

However, I know your means now, we can provide `cleanTcs` method in `ctrl/client`.

Reply:
Done